File: packages/typescript/src/client/http/TokenManager.ts
import type { AuthConfig } from "./types";

/** Manages authentication tokens for API requests */
export class TokenManager {
  private staticToken: string | undefined;
  private getTokenFn: (() => Promise<string> | string) | undefined;

  constructor(auth?: AuthConfig) {
    if (auth?.type === "static") {
      this.staticToken = auth.token;
    } else if (auth?.type === "dynamic") {
      this.getTokenFn = auth.getToken;
    }
  }

  /** Get the current token (async to support dynamic tokens) */
  async getToken(): Promise<string | undefined> {
    if (this.staticToken) {
      return this.staticToken;
    }
    if (this.getTokenFn) {
      return await this.getTokenFn();
    }
    return undefined;
  }

  /** Update static token (for token refresh scenarios) */
  setToken(token: string): void {
    this.staticToken = token;
    this.getTokenFn = undefined;
  }

  /** Check if authentication is configured */
  hasAuth(): boolean {
    return !!(this.staticToken || this.getTokenFn);
  }
}

File: packages/csharp/src/PokeForge.SDK/Client/Http/TokenManager.cs
namespace PokeForge.SDK.Client.Http;

/// <summary>
/// Manages authentication tokens for API requests.
/// Supports both static tokens and dynamic token providers.
/// </summary>
internal sealed class TokenManager
{
    private string? _staticToken;
    private Func<CancellationToken, Task<string>>? _tokenProvider;

    /// <summary>
    /// Creates a new TokenManager with optional static token and/or dynamic token provider.
    /// </summary>
    /// <param name="staticToken">A static JWT token.</param>
    /// <param name="tokenProvider">A function that returns a fresh token when called.</param>
    public TokenManager(string? staticToken, Func<CancellationToken, Task<string>>? tokenProvider)
    {
        _staticToken = staticToken;
        _tokenProvider = tokenProvider;
    }

    /// <summary>
    /// Gets the current token. Returns null if no authentication is configured.
    /// If a token provider is configured, it takes precedence over the static token.
    /// </summary>
    public async ValueTask<string?> GetTokenAsync(CancellationToken cancellationToken = default)
    {
        if (_tokenProvider is not null)
        {
            return await _tokenProvider(cancellationToken).ConfigureAwait(false);
        }

        return _staticToken;
    }

    /// <summary>
    /// Updates the static token. Clears any token provider.
    /// Useful for token refresh scenarios.
    /// </summary>
    public void SetToken(string token)
    {
        _staticToken = token;
        _tokenProvider = null;
    }

    /// <summary>
    /// Sets a new token provider. Clears any static token.
    /// </summary>
    public void SetTokenProvider(Func<CancellationToken, Task<string>> tokenProvider)
    {
        _staticToken = null;
        _tokenProvider = tokenProvider;
    }

    /// <summary>
    /// Returns true if any form of authentication is configured.
    /// </summary>
    public bool HasAuth => _staticToken is not null || _tokenProvider is not null;
}
